include/OpenSpaceToolkit/Astrodynamics/Solvers/FiniteDifferenceSolver.hpp

```diff
@@ -0,0 +1,137 @@
+#include <OpenSpaceToolkit/Physics/Time/Instant.hpp>
+
+#include <OpenSpaceToolkit/Astrodynamics/Trajectory/State.hpp>
+
+namespace ostk
+{
+namespace astro
+{
+namespace solvers
+{
+
+using ostk::core::types::Real;
+using ostk::core::types::String;
+using ostk::core::ctnr::Array;
+
+using ostk::math::obj::VectorXd;
+using ostk::math::obj::MatrixXd;
+
+using ostk::physics::time::Instant;
+using ostk::physics::time::Duration;
+
+using ostk::astro::trajectory::State;
+
+/// @brief                      Finite Difference solver
+
+class FiniteDifferenceSolver
+{
+   public:
+    enum class Type
+    {
+        Forward,
+        Backward,
+        Central
+    };
+
+    /// @brief                  Constructor
+    ///
+    /// @code
+    ///                         const Real stepPercentage = 1e-3;
+    ///
+    ///                         FiniteDifferenceSolver finiteDifferenceSolver = {stepPercentage};
+    ///
+    /// @endcode
+    ///
+    /// @param                  [in] aType A Finite Difference type.
+
+    FiniteDifferenceSolver(const Type& aType);
+
+    /// @brief                  Output stream operator.
+    ///
+    /// @param                  [in] anOutputStream An output stream.
+    /// @param                  [in] aFiniteDifference A finite difference solver.
+    /// @return                 An output stream.
+
+    friend std::ostream& operator<<(std::ostream& anOutputStream, const FiniteDifferenceSolver& aFiniteDifference);
+
+    /// @brief                  Compute the State Transition Matrix by perturbing the coordinates
+    ///
+    /// @param                  [in] aState A state.
+    /// @param                  [in] anInstantArray An array of instants.
+    /// @param                  [in] getStates Callable to generate States at the requested Instants.
+    /// @param                  [in] aStepPercentage The step percentage to use for the perturbation. Defaults to 1e-3.
+    ///
+    /// @return                 The State Transition Matrix
+
+    MatrixXd computeStateTransitionMatrix(
+        const State& aState,
+        const Array<Instant>& anInstantArray,
+        std::function<Array<State>(const State&, const Array<Instant>&)> getStates,
+        const Real& aStepPercentage = 1e-3
+    ) const;
+
+    /// @brief                  Compute the State Transition Matrix by perturbing the coordinates
+    ///
+    /// @param                  [in] aState A state.
+    /// @param                  [in] anInstant An instant.
+    /// @param                  [in] getState Callable to generate a State at the requested Instant.
+    /// @param                  [in] aStepPercentage The step percentage to use for the perturbation. Defaults to 1e-3.
+    ///
+    /// @return                 The State Transition Matrix
+
+    MatrixXd computeStateTransitionMatrix(
+        const State& aState,
+        const Instant& anInstant,
+        std::function<State(const State&, const Instant&)> getState,
```

Comment:
```suggestion
        std::function<State(const State&, const Instant&)> propagateState,
```

nit: a bit pedantic, but something like `propagateState` (or `getStateAtInstant` or `propagateStateToInstant`) is a little clearer IMO and describes what the input lambda should do.

Reply:
I've updated this to generateState, agreed that getState is a bit generic and used as getter.